feat(topology): Topology plugin visualisation of workload

invincibleJai opened the pull request:
**Tracks:** https://github.com/janus-idp/backstage-plugins/issues/58

**Fixes:**  https://github.com/janus-idp/backstage-plugins/issues/85, https://github.com/janus-idp/backstage-plugins/issues/89, https://github.com/janus-idp/backstage-plugins/issues/105

**Screenshots/GIF's:**
- Light Theme
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/5129024/219265140-a56977b1-6a85-433b-886b-ff187c54c824.png">


- Dark Theme
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/5129024/219265235-04f7900a-a0e2-46d1-b2a2-bd807fe6ca2a.png">


- Empty State
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/5129024/218790057-56df8c8e-1cad-45a4-9244-b642ae5ceb5d.png">

**Unit Tests:**
<img width="732" alt="image" src="https://user-images.githubusercontent.com/5129024/220941872-73a0ee70-0f6e-42c8-b553-0fd2b4df8521.png">



**Test Setup:**
Install and configure Kubernetes plugin by following these guides:

https://backstage.io/docs/features/kubernetes/installation
https://backstage.io/docs/features/kubernetes/configuration

Kubernetes plugin is properly configured and able to connect to the cluster via a ServiceAccount

Following ClusterRole must be granted to ServiceAccount accessing the cluster:
(k8s plugin by default tries to fetch the following resources and if access is not provided an error is shown)

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: backstage-read-only
rules:
  - apiGroups:
      - '*'
    resources:
      - pods
      - configmaps
      - services
      - deployments
      - replicasets
      - horizontalpodautoscalers
      - ingresses
      - statefulsets
      - limitranges
      - daemonsets
    verbs:
      - get
      - list
      - watch
  - apiGroups:
      - batch
    resources:
      - jobs
      - cronjobs
    verbs:
      - get
      - list
      - watch
  - apiGroups:
      - metrics.k8s.io
    resources:
      - pods
    verbs:
      - get
      - list
 ```
 
Create a service component entity - https://github.com/temp-demo-organisation/nationalparks-py/blob/master/catalog-info2.yaml

To get the resources from a k8s cluster do the following:

K8s plugin identifies if the provided entity has k8s resources and if yes from which namespace it should get the resources based on the following annotations added to the entity's catalog-info.yaml
```
annotations: 
 backstage.io/kubernetes-id: nationalparks-py
 backstage.io/kubernetes-namespace: div
```

For k8s plugin to get the desired entity's k8s resources the following label should be added to the resources `'backstage.io/kubernetes-id': nationalparks-py`

A custom label selector can also be added which will then be used by Backstage to find the resources. The label selector takes precedence over the id annotation.
`'backstage.io/kubernetes-label-selector': 'app=my-app,component=front-end`

In order to use k8s plugin to get the resources for Topology view, the above annotations/labels should be in place.

Enable additional tab on the entity view page

   ```ts
   // packages/app/src/components/catalog/EntityPage.tsx
   import { TopologyPage } from '@janus-idp/backstage-plugin-topology';

   const serviceEntityPage = (
     <EntityPageLayout>
       // ...
       <EntityLayout.Route path="/topology" title="Topology">
        <TopologyPage />
       </EntityLayout.Route>
     </EntityPageLayout>
   );
   ```

**Todo:**
- [x] Visualize resources of the Deployment kind as read-only
- [x] Visualize app group for deployments if present with the label `'[app.kubernetes.io/part-of]': <GROUP_NAME>`
- [x] Visualize a visual connecter between two deployments if present with annotation `'[app.openshift.io/connects-to]': '[{"apiVersion": <RESOURCE_APIVERSION>,"kind": <RESOURCE_KIND>,"name": <RESOURCE_NAME>}]'`
- [x] Light/Dark theme support
- [x] Integrate with the k8s plugin to communicate with the cluster
- [x] Added Empty state for topology if no resources exist
- [x] Clean up of code/add tests

tumido commented on plugins/topology/src/components/Topology/utils/resource-utils.ts:
```diff
@@ -0,0 +1,104 @@
+
+export const MONITORABLE_KINDS = [
+  'Deployment',
+  'DeploymentConfig',
```
What happens if the plugin is unable to locate resource type on the cluster? Does it raise errors/exceptions? I haven't run the plugin against vanilla Kubernetes myself, I'm just curious. I'm asking on the basis that `DeploymentConfig`s are not Kubernetes resources.

invincibleJai replied:
yeah, have removed DeploymentConfig and this is the initial PR here we are looking for `'deployments', 'pods', 'services', 'ingresses', 'replicasets'` and node workload is formed for Deployment, stadalone Pod others resources are needed to form relationship i.e pods and replicasets to show deployment. There would be follow-up PRs to add a decorator for ingress on workload and sidebar which will open on click to show other associated resources, labels, annotations, etc

tumido reviewed (commented):
I've left few suggestions inline.

I would also suggest to rewrite the use of `lodash` methods into native JS (like those `_.map`s, `_.size`s, `_.head`s, etc...), WDYT?

invincibleJai commented on packages/app/src/components/catalog/EntityPage.tsx:
```diff
@@ -192,6 +193,11 @@ const websiteEntityPage = (
+    {/* TODO : Remove for testing Add entity for topology */}
+    <EntityLayout.Route path="/topology" title="Topology">
+      <TopologyPage />
+    </EntityLayout.Route>
```
Thanks updated.

invincibleJai commented on plugins/topology/src/components/Topology/types/topology-types.ts:
```diff
@@ -0,0 +1,51 @@
+import { K8sResourceKind, WatchK8sResults } from './types';
+
+export type TrafficData = {
+  nodes: KialiNode[];
```
I have removed it(moved types from the console, and cleaned it up now.

divyanshiGupta commented:
@invincibleJai On changing tabs and then coming back to topology, empty state is being shown first even though the node data is available. It might be because it takes time to load the data and then render the topology view. Instead of empty state in this case we can show a loading visualisation similar to ocp.


https://user-images.githubusercontent.com/20724543/219331295-cf8f13e4-b68f-4bd0-a619-34c4ff4d5ca2.mov

tumido commented on plugins/topology/README.md:
```diff
@@ -0,0 +1,107 @@
+   https://backstage.io/docs/features/kubernetes/installation
+   https://backstage.io/docs/features/kubernetes/configuration
+2. Kubernetes plugin is properly configured and able to connect to the cluster via a `ServiceAccount`
+3. Following `ClusterRole` must be granted to ServiceAccount accessing the cluster: (k8s plugin by default tries to fetch the following resources and if access is not provided an error is shown)
```
I think we should mention this that this role is already granted if you have the Backstage Kubernetes Plugin properly configured (which is a prerequisite), and link to https://backstage.io/docs/features/kubernetes/configuration#role-based-access-control

The current formulation gives impression that there is something different in the role itself when compared to the one required by kubernetes plugin. While that may become the case in future, I think we should rather formulate in a way that basically says: "Please grant the permissions that Backstage Kubernetes plugin requires + these additional ones" (once we have any additional ones).

tumido commented on plugins/topology/src/components/Graph/TopologyComponentFactory.tsx:
```diff
@@ -0,0 +1,35 @@
+    return withPanZoom()(DefaultGraph);
+  }
+  switch (type) {
+    case 'workload':
```
This should be a const as well (it't used in [`data-transforms/data-transformer.ts`](https://github.com/janus-idp/backstage-plugins/pull/104/files#diff-0253d70e4353e47b4ecae9126d55b0479178517ec598e1861b889c9dd04cd53fR19) as well)

tumido commented on plugins/topology/src/const.ts:
```diff
@@ -0,0 +1,20 @@
+export const TYPE_APPLICATION_GROUP = 'part-of';
+export const TYPE_CONNECTS_TO = 'connects-to';
+
+export const TYPE_TRAFFIC_CONNECTOR = 'traffic-connector';
```
This is not used.

tumido commented on plugins/topology/src/utils/topology-utils.ts:
```diff
@@ -0,0 +1,134 @@
+    },
+    data: {
+      kind: kindRef,
+      editURL: deploymentsAnnotations['app.openshift.io/edit-url'],
+      vcsURI: deploymentsAnnotations['app.openshift.io/vcs-uri'],
+      vcsRef: deploymentsAnnotations['app.openshift.io/vcs-ref'],
```
Can we document these annotations?

invincibleJai replied:
Have removed this now. These annotations are in the console we had for showing the repo and on click user can directly go to that but we'll bring this as a decorator for workload like one for ingress.

tumido commented on plugins/topology/src/utils/transform-utils.ts:
```diff
@@ -0,0 +1,266 @@
+): OdcNodeModel => {
+  const uid = resource.metadata?.uid;
+  const name = resource.metadata?.name;
+  const label = resource.metadata?.labels?.['app.openshift.io/instance'];
```
Shouldn't this be

```suggestion
  const label = resource.metadata?.labels?.['app.kubernetes.io/instance'];
```

At least that's what is used at [`utils/topology-utils.ts`](https://github.com/janus-idp/backstage-plugins/pull/104/files#diff-8eed6f305ecf78ad964f1776378d9f7c65b816225105701acbf82ff4773cc8a2R96) above.

Maybe the `app.kubernetes.io/instance` should be a proper global constant.

tumido commented on plugins/topology/src/utils/transform-utils.ts:
```diff
@@ -0,0 +1,266 @@
+  annotations: any,
+): (string | ConnectsToData)[] => {
+  let edges: (string | ConnectsToData)[] = [];
+  if (_.has(annotations, ['app.openshift.io/connects-to'])) {
```
`app.openshift.io/connects-to` should be a global constant.

tumido reviewed (changes requested):
I've just skimmed the code. It's a lot of content to digest for a thorough review.

I see quite some inconsistent constant usage, can we globalize that and maybe document some annotations that user can define but they are hidden in code?

See my comments inline

riginoommen commented on plugins/topology/README.md:
```diff
@@ -0,0 +1,107 @@
+# Topology plugin for Backstage
+
+This plugin will help with visualizing the workloads like Deployment, Replicasets, Pods powering any serivce on kubernets cluster.
```
```suggestion
This plugin will help with visualizing the workloads like Deployment, Replicaset and Pods powering any service on the Kubernetes cluster.
```

riginoommen commented on plugins/topology/README.md:
```diff
@@ -0,0 +1,107 @@
+
+## Prerequisites
+
+1. Install and configure kubernetes plugin by following these guides:
```
```suggestion
1. Install and configure Kubernetes plugin by following these guides:
```

riginoommen commented on plugins/topology/src/hooks/useDeepCompareMemoize.ts:
```diff
@@ -0,0 +1,19 @@
+
+export const useDeepCompareMemoize = <T = any>(
+  value: T,
+  strinfigy?: boolean,
```
Looks like type for ```stringify```

riginoommen commented on plugins/topology/src/hooks/useDeepCompareMemoize.ts:
```diff
@@ -0,0 +1,19 @@
+  const ref = React.useRef<T>();
+
+  if (
+    strinfigy
```
Looks like type for ```stringify```

riginoommen commented on plugins/topology/src/utils/pod-resource-utils.ts:
```diff
@@ -0,0 +1,280 @@
+export const CONTAINER_WAITING_STATE_ERROR_REASONS = [
+  'CrashLoopBackOff',
+  'ErrImagePull',
+  'ImagePullBackOff',
```
```suggestion
  'ImagePullBackOff',
  'InvalidImageName'
  'CreateContainerError'
```

riginoommen commented on plugins/topology/src/utils/pod-resource-utils.ts:
```diff
@@ -0,0 +1,280 @@
+import * as _ from 'lodash';
```
I suggest importing the specific resources which is utilized in the source code.

```suggestion
import { get, isFinite, some } from 'lodash';
```

riginoommen commented on plugins/topology/src/utils/resource-utils.ts:
```diff
@@ -0,0 +1,40 @@
+import * as _ from 'lodash';
```
I suggest importing the specific resources which is utilized in the source code.

```suggestion
import { isEmpty} from 'lodash';
```

riginoommen commented on plugins/topology/src/utils/resource-utils.ts:
```diff
@@ -0,0 +1,40 @@
+};
+
+const isStandaloneJob = (job: K8sResourceKind) =>
+  !_.find(job.metadata?.ownerReferences, owner => owner.kind === 'CronJob');
```
Can we utilize the find method in the javascript instead of the library

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

riginoommen commented on plugins/topology/src/utils/resource-utils.ts:
```diff
@@ -0,0 +1,40 @@
+export const validPod = (pod: K8sResourceKind) => {
+  const owners = pod?.metadata?.ownerReferences;
+  const phase = pod?.status?.phase;
+  return _.isEmpty(owners) && phase !== 'Succeeded' && phase !== 'Failed';
```
can we try utiizing the ```Object.keys()``` method to identify the empty object

riginoommen commented on plugins/topology/src/utils/topology-utils.ts:
```diff
@@ -0,0 +1,134 @@
+import * as _ from 'lodash';
```
```suggestion
import { flatten, get } from 'lodash';
```

riginoommen commented on plugins/topology/src/utils/transform-utils.ts:
```diff
@@ -0,0 +1,266 @@
+  NodeModel,
+  NodeShape,
+} from '@patternfly/react-topology';
+import * as _ from 'lodash';
```
```suggestion
import { get} from 'lodash';
```

invincibleJai replied:
Thanks, am working on refactoring will replace lodash usage with native methods wherever applicable and prefer named import over all as suggeted

riginoommen reviewed (changes requested):
I suggest to iupdate the imports and use native code for the better performance

riginoommen commented on plugins/topology/README.md:
````diff
@@ -0,0 +1,107 @@
+1. Install the plugin
+
+   ```bash
+   yarn workspace app add @janus-idp/backstage-plugin-topology
````
```suggestion
   yarn add --cwd packages/app @janus-idp/backstage-plugin-topology
```

tumido replied:
Is there a reason why this should be preferred? Can you please elaborate?

riginoommen replied:
For most of the upstream packages for the backstage this format is commonly used to make sync on this I suggested

riginoommen commented on plugins/topology/dev/index.tsx:
```diff
@@ -0,0 +1,115 @@
+  kubernetesApiRef,
+} from '@backstage/plugin-kubernetes';
+import { TestApiProvider } from '@backstage/test-utils';
+import fixture1 from '../src/__fixtures__/1-deployments.json';
```
```fixture1``` variable seems odd i suggest using a variable name which is little more understandable

riginoommen commented on plugins/topology/package.json:
```diff
@@ -0,0 +1,66 @@
+    "@testing-library/jest-dom": "^5.10.1",
+    "@testing-library/react": "^12.1.3",
+    "@testing-library/user-event": "^14.0.0",
+    "@types/node": "*",
```
Can we mention a version and lock it based on the compatiblity

riginoommen commented on plugins/topology/src/components/Graph/WorkloadNode.tsx:
```diff
@@ -0,0 +1,194 @@
+    const [hover, hoverRef] = useHover();
+    const size = Math.min(width, height);
+    const { radius, decoratorRadius } = calculateRadius(size);
+    const cx = width / 2;
+    const cy = height / 2;
```
```suggestion
    const cx = width / 2;
    const cy = height / 2;
```

can you please use a variable which is more readable

riginoommen commented on plugins/topology/src/components/Pods/pod.ts:
```diff
@@ -0,0 +1,52 @@
+import { global_palette_white as globalWhite } from '@patternfly/react-tokens/dist/js/global_palette_white';
+import { global_warning_color_100 as globalWarning100 } from '@patternfly/react-tokens/dist/js/global_warning_color_100';
+
+export enum AllPodStatus {
```
```suggestion
export enum AllPodStatus {
```

suggest to make the enums in a common format like ALL_POD_STATUS

invincibleJai replied:
shouldn't enums be in PascalCase i.e UpperCamelCase , it was not consistent across, and have updated those

https://www.typescriptlang.org/docs/handbook/enums.html#:~:text=Enums%20are%20one%20of%20the,numeric%20and%20string%2Dbased%20enums.

riginoommen commented on plugins/topology/src/const.ts:
```diff
@@ -0,0 +1,20 @@
+export const DEFAULT_NODE_PAD = 20;
```
Differentiating the ENUm declaration standard and constant declaration standard helps to make code more readable

invincibleJai replied:
sorry if I get it right all enums I need to update to PascalCase i.e UpperCamelCase and constants should be as it's here or ? 

https://www.typescriptlang.org/docs/handbook/enums.html#:~:text=Enums%20are%20one%20of%20the,numeric%20and%20string%2Dbased%20enums.

riginoommen commented on plugins/topology/src/hooks/useForceUpdate.ts:
```diff
@@ -0,0 +1,9 @@
+import * as React from 'react';
+
+export const useForceUpdate = () => {
+  const [, setTick] = React.useState(0);
```
```suggestion
  const [ setTick] = React.useState(0);
```

invincibleJai replied:
am not sure if the above suggestion will work as `useState` returns a tuple with the first position will return an item and the second setter for it

spadgett commented:
@invincibleJai How much of this code was taken from the OpenShift console topology? Should we think about publishing common code as a package to avoid duplication?

invincibleJai commented:
> @invincibleJai How much of this code was taken from the OpenShift console topology? Should we think about publishing common code as a package to avoid duplication?

@spadgett graph shapes like WorkloadNode, BaseNode, and PodSet have been used and associated utils and types have been pulled from OpenShift console but that also involved a lot of refactoring as Console code has a dependency on a lot of internal / shared packages (some interdependency with devConsole package as well).

In backstage plugin, the initial version is read-only and should work with any k8s cluster.

Moving it to a separate package needs to be evaluated and this will be a lot more work and needs to discuss the strategy for updates as well.

invincibleJai commented:
> @invincibleJai On changing tabs and then coming back to topology, empty state is being shown first even though the node data is available. It might be because it takes time to load the data and then render the topology view. Instead of empty state in this case we can show a loading visualisation similar to ocp.

Thanks @divyanshiGupta, have fixed this.

spadgett commented on plugins/topology/src/types/types.ts:
```diff
@@ -0,0 +1,133 @@
+
+// Generic, unknown kind. Avoid when possible since it allows any key in spec
+// or status, weakening type checking.
+export type K8sResourceKind = K8sResourceCommon & {
```
We're trying to slowly remove this from OpenShift console, so would be good to avoid using it here if we can. It's the `[key: string]: any` which makes the type checking much weaker even if you extend this type because you don't catch any typos in property names.

invincibleJai replied:
so shall I add strict types for supported resources over generic ones like K8sDeploymentKind, K8sPodKind, etc?

invincibleJai replied:
i checked backstage-k8s plugin  and it fetches native types from https://www.npmjs.com/package/@kubernetes/client-node , have used the same to be more strict

https://github.com/backstage/backstage/blob/master/plugins/kubernetes/package.json#L43
https://github.com/backstage/backstage/blob/master/plugins/kubernetes/src/components/CronJobsAccordions/CronJobsAccordions.tsx#L25

spadgett commented on plugins/topology/src/utils/topology-utils.ts:
```diff
@@ -0,0 +1,116 @@
+  return topologyObject.resource || topologyObject.resources?.obj;
+};
+
+const isGroupVersionKind = (ref: string) => ref.split('~').length === 3;
```
How are the `~` delimited reference strings used in topology? This is another OpenShift console-ism that it might be good to remove unless we need it.

spadgett reviewed (commented):
Thanks, @invincibleJai. Fantastic progress. I know this is a huge effort.

I'm not super familiar with topology, but I believe @jeff-phillips-18 has looked as well?

I think we should prioritize writing some tests. If we don't hold the initial PR on it, at least after it goes in.

invincibleJai commented on plugins/topology/README.md:
```diff
@@ -0,0 +1,67 @@
+1. Install and configure Kubernetes plugin by following these guides:
+   https://backstage.io/docs/features/kubernetes/installation
+   https://backstage.io/docs/features/kubernetes/configuration
+2. Kubernetes plugin is properly configured and able to connect to the cluster via a `ServiceAccount`
```
It should be any supported authentication mechanism as provided here https://backstage.io/docs/features/kubernetes/configuration 

and has cluster roles for https://backstage.io/docs/features/kubernetes/configuration#role-based-access-control 

we have added `ServiceAccount` as that's what we have tested. Once we verify other mechanisms will update.

@divyanshiGupta ^^

invincibleJai commented on plugins/topology/src/components/Topology/TopologyComponent.tsx:
```diff
@@ -0,0 +1,46 @@
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [type]);
+
+  const watchedResources = [
```
Yeah, supported default types from k8s plugin is listed here https://backstage.io/docs/features/kubernetes/configuration#objecttypes-optional

Yeah eventually we would need to bring in support for custom resources as well

invincibleJai commented on plugins/topology/src/components/Graph/WorkloadNode.tsx:
```diff
@@ -0,0 +1,194 @@
+const getNodePodStatus = (podStatus: AllPodStatus): number => {
+  switch (podStatus) {
+    case AllPodStatus.Failed:
+    case AllPodStatus.CrashLoopBackOff:
```
Have added `ErrImagePull` as well here. 

<img width="355" alt="image" src="https://user-images.githubusercontent.com/5129024/220551351-b20e5502-5860-47b9-8c60-5a7b55805bd9.png">


In podStatus for topology I don't see this being handled in the console as there in topology, error in podrings will be shown only pod status is failed and for errImagePull status will be in Pending

```
...
status:
  phase: Pending
  ...
  containerStatuses:
    - name: container
      state:
        waiting:
          reason: ImagePullBackOff
          message: Back-off pulling image "jai/helloworld56"

```
and the best reasoning I can think of is there we have git import flows where we do s2i or pipeline builds so until the build is complete the image won't be accessible so the user will see a pending state over failed.

spadgett replied:
There are a number of ways this can occur (typo in image name, private image that can't be pulled, image registry is down, etc). We probably want to handle `ImagePullBackOff` as well.

spadgett replied:
Yeah, I guess I can see the argument that the build hasn't completed yet, though, in which case we want pending. Maybe we should check to see if it wasn't intentional.

invincibleJai replied:
currently, it'll show pending only if phase is pending with the exception to `CrashLoopBackOff` ,  `ImagePullBackOff ` or state is terminated with `exitCode != 0` in containerStatuses state

 https://github.com/janus-idp/backstage-plugins/pull/104/files#diff-8da95685163a1fdc3228145eccfedce06be50826d3d5f2e64d135106c5a02b21R65-R76

invincibleJai commented on plugins/topology/src/components/Topology/TopologyWorkloadView.tsx:
```diff
@@ -0,0 +1,113 @@
+      };
+      controller.fromModel(model, false);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
not really, didn't observe any updated with `controller` in the dependency array.

invincibleJai commented on plugins/topology/src/models.ts:
```diff
@@ -0,0 +1,54 @@
+import { K8sModel } from './types/types';
+
+export const ReplicaSetModel: K8sModel = {
```
Have updated it to use group/version/kind and later if we feel the need can add more props like custom Abbr/color if needed for badges in topology.

jeff-phillips-18 commented on plugins/topology/src/components/Graph/BaseNode.tsx:
```diff
@@ -0,0 +1,126 @@
+  const [hover, internalHoverRef] = useHover();
+  const nodeHoverRefs = useCombineRefs(
+    internalHoverRef,
+    hoverRef as React.Ref<Element>,
```
There are some know type issues in the PF topology package. These are planned to be fixed in the next major release.

jeff-phillips-18 commented on plugins/topology/src/components/Graph/BaseNode.tsx:
```diff
@@ -0,0 +1,126 @@
+  element: Node;
+  hoverRef?: (node: Element) => () => void;
+  dragging?: boolean;
+  dropTarget?: boolean;
+  canDrop?: boolean;
```
Don't need these.

jeff-phillips-18 commented on plugins/topology/src/components/Graph/BaseNode.tsx:
```diff
@@ -0,0 +1,126 @@
+  alertVariant?: NodeStatus;
+} & Partial<WithSelectionProps> &
+  Partial<WithDragNodeProps> &
+  Partial<WithDndDropProps> &
+  Partial<WithContextMenuProps> &
+  Partial<WithCreateConnectorProps>;
```
```suggestion
  Partial<WithContextMenuProps>;
```
We don't need these props if we aren't allowing edge modifications.

invincibleJai commented:
> Thanks, @invincibleJai. Fantastic progress. I know this is a huge effort.
> 
> I'm not super familiar with topology, but I believe @jeff-phillips-18 has looked as well?
> 
> I think we should prioritize writing some tests. If we don't hold the initial PR on it, at least after it goes in.

Thanks @spadgett for reviewing, have addressed most of the review comments and have added tests as well for some of the components (will continue adding more and it can be a follow-up PR as well).

spadgett reviewed (commented):
Overall, looks good. Just a few small comments.

Thank you for the tests!

invincibleJai commented on plugins/topology/src/models.ts:
```diff
@@ -0,0 +1,37 @@
+  kind: 'Service',
+};
+
+export enum ModelsPlural {
```
updated model to just be `GroupVersionKind` so have kept it separate.

invincibleJai commented on plugins/topology/src/types/types.ts:
```diff
@@ -0,0 +1,40 @@
+  [key: string]: any;
+};
+
+type ContainerState = {
```
Thanks 👍 

yeah we do have `V1ContainerStatus` itself, updated to use those.

spadgett commented on plugins/topology/src/models.ts:
```diff
@@ -0,0 +1,37 @@
+import { GroupVersionKind } from './types/types';
+
+export const ReplicaSetModel: GroupVersionKind = {
```
nitpick: maybe `ReplicaSetGroupVersionKind` rather than `Model`.

spadgett reviewed (commented):
LGTM, just two small naming nits

invincibleJai commented on plugins/topology/src/types/types.ts:
```diff
@@ -0,0 +1,18 @@
+  apiGroup?: string;
+};
+
+export type K8sResourceKind = V1Deployment | V1Pod | V1Service | V1ReplicaSet;
```
will call it `K8sWorkloadResource` as these also include service and ingress in next PR which may not be workload but workload resources

invincibleJai commented:
Thanks @spadgett , @jeff-phillips-18 for the review 👍

invincibleJai commented:
@tumido PTAL when you get time, thanks!!

tumido commented on plugins/topology/package.json:
```diff
@@ -0,0 +1,69 @@
+  "main": "src/index.ts",
+  "types": "src/index.ts",
+  "license": "Apache-2.0",
+  "private": true,
```
```suggestion
```

Please make the package public so we can release it.

invincibleJai replied:
Thanks Tom, updated

tumido reviewed (approved):
/lgtm
/approve

This is a huge PR. kudos to everyone working on it and reviewing it. :slightly_smiling_face: :100: Good job!

@spadgett can you please take one last look? If you approve we can merge and release it :slightly_smiling_face:

spadgett reviewed (approved):
LGTM 👍
